src/Core/Containers/VectorArray.hpp

```diff
@@ -2,92 +2,93 @@
     /// Inheriting constructors from std::vector
     using AlignedStdVector<V>::AlignedStdVector;
 
-    /// Returns the array as an Eigen Matrix Map
-    MatrixMap getMap() {
-        CORE_ASSERT( !this->empty(), "Cannot map an empty vector " );
-        return MatrixMap( this->data()->data(), V::RowsAtCompileTime, this->size() );
-    }
-
-    /// Returns the array as an Eigen Matrix Map (const version)
-    ConstMatrixMap getMap() const {
-        CORE_ASSERT( !this->empty(), "Cannot map an empty vector " );
-        return ConstMatrixMap( this->data()->data(), V::RowsAtCompileTime, this->size() );
-    }
-};
-
-/// This specialization stores an array of scalars which can be used as a dynamic
-/// Eigen column vector.
-template <>
-class VectorArray<float> : public AlignedStdVector<float>
-{
-  public:
-    // Type shortcuts
-    using Matrix         = Eigen::Matrix<float, 1, Eigen::Dynamic>;
-    using MatrixMap      = Eigen::Map<Matrix>;
-    using ConstMatrixMap = Eigen::Map<const Matrix>;
-
-  public:
-    /// Inheriting constructors from std::vector
-    using AlignedStdVector<float>::AlignedStdVector;
+    /// @{
+    /// ContainerIntrosectionInterface implementation
+    size_t getSize() const override { return this->size(); }
+    size_t getNumberOfComponents() const override { return std::max( 0, NumberOfComponents ); }
+    size_t getBufferSize() const override { return getSize() * sizeof( V ); }
+    int getStride() const override { return sizeof( V ); }
+    const void* dataPtr() const override { return this->data(); }
+    /// @}
 
     /// Returns the array as an Eigen Matrix Map
     MatrixMap getMap() {
         CORE_ASSERT( !this->empty(), "Cannot map an empty vector " );
-        return MatrixMap( this->data(), 1, Eigen::Index( this->size() ) );
+        return MatrixMap( TypeHelper::getData( this ),
+                          TypeHelper::NumberOfComponents,
+                          Eigen::Index( this->size() ) );
     }
 
     /// Returns the array as an Eigen Matrix Map (const version)
     ConstMatrixMap getMap() const {
         CORE_ASSERT( !this->empty(), "Cannot map an empty vector " );
-        return ConstMatrixMap( this->data(), 1, Eigen::Index( this->size() ) );
+        return MatrixMap( TypeHelper::getConstData( this ),
+                          TypeHelper::NumberOfComponents,
+                          Eigen::Index( this->size() ) );
     }
 };
 
-template <>
-class VectorArray<double> : public AlignedStdVector<double>
-{
-  public:
-    // Type shortcuts
-    using Matrix         = Eigen::Matrix<double, 1, Eigen::Dynamic>;
-    using MatrixMap      = Eigen::Map<Matrix>;
-    using ConstMatrixMap = Eigen::Map<const Matrix>;
-
-  public:
-    /// Inheriting constructors from std::vector
-    using AlignedStdVector<double>::AlignedStdVector;
+template <typename V>
+struct VectorArrayTypeHelper<V, true, false> {
+    using component_type                    = V;
+    static constexpr int NumberOfComponents = 1;
```

Comment:
if using `std::is_scalar<V>::value` above, should use `(std::is_arithmetic<T>::value || std::is_enum<T>::value) ? 1 : 0` here